QoL Dispenser Update

Protonull opened the pull request:
This change has been made [at the request of DriftL0rd](https://www.reddit.com/r/civclassics/comments/cq34as/minor_suggestion_regarding_xp_bottles/ewu238l/?context=6), he noticed that xp can only be granted via players throwing xp bottles after creating a machine to auto dispense. The response from the community was to install mods, I object to the line of thinking that gameplay should be replaced with mods, and so this change gives players more freedom. It also fixes a funny consequence of orb xp being disabled, see the individual commit messages for more information.

Protonull committed:
No More EXP Littering

When `block_natural_exp` is enabled, xp orbs will now be teleported to the player's location as they have no effect so there's no reason to let them linger where the xp bottle was thrown.

Protonull committed:
Allows Dispensed XP

This update allows for players to receive dispensed xp even when `block_natural_exp` is enabled. Instead of littering the ground with xp orbs that grant xp when picked up, it instead mimics the behaviour of player thrown bottles whereby the xp is granted instantly regardless of orbs. When bottle is dispensed and hits something, it finds the nearest player within a block radius of the location, the closest player receives the xp instantly. If no player is found then the xp is lost.

Protonull committed:
Bump version (1.0.1)

Maxopoly commented on src/main/java/com/civclassic/oldenchanting/OldEnchanting.java:
```diff
@@ -302,13 +306,41 @@ private void createXPBottles(Player player, int totalExp) {
+			if (source instanceof Player) {
 				Player shooter = (Player) source;
 				shooter.giveExp(xpPerBottle);
+				bottle.teleport(shooter.getEyeLocation());
```
Why is this needed

Protonull replied:
When players throw Bottles o' Enchanting, they receive the xp on impact, regardless of the impact's location relative to the player. It means that the orbs themselves play no part in the player's leveling up, so the orbs will linger at the impact site until they expire or picked up.

This change teleports the orbs to the player at the moment of impact and in doing so removing them from the game but also providing the player with an audible feedback for receiving the xp.

Maxopoly commented on src/main/java/com/civclassic/oldenchanting/OldEnchanting.java:
```diff
@@ -302,13 +306,41 @@ private void createXPBottles(Player player, int totalExp) {
+				bottle.teleport(shooter.getEyeLocation());
+			}
+			else if (source instanceof BlockProjectileSource) {
+				List<Player> nearby = bottle.getNearbyEntities(0.9, 0.9, 0.9)
```
Could easily make that bigger to something like 5 blocks, ideally it'd even be a config parameter

Protonull replied:
Done, the radius is now a config option with the recommended settings being in between two and seven

Maxopoly commented on src/main/java/com/civclassic/oldenchanting/OldEnchanting.java:
```diff
@@ -302,13 +306,41 @@ private void createXPBottles(Player player, int totalExp) {
+				}
+				Player closest = null;
+				double distance = Double.MAX_VALUE;
+				for (Player player : nearby) {
```
If you are already doing fancy stream stuff, you might as well go all the way and do
```
.map((entity) -> (Player) entity)
.min(p ->  p.getLocation().distanceSquared(bottle.getLocation()))
```

which gives you an `Optional<Player>`, circumventing the need for both the collector and your for loop.

Maxopoly replied:
In that case you could even do some obvious particle effect if the Optional is null to indicate that the XP is being wasted right now

Protonull committed:
Post-suggestion Update

1. The dispensed xp detection radius is now a config option with a recommended setting between two and seven.
2. Orbs from bottles thrown from players will be teleported to the player's feet instead of their head as there's no need to blind the player with orbs.
3. Slightly changed the distance location to take the advantage away from the high ground as player location's are calculated from their feet.
4. Orbs from dispensers will be teleported to the player, otherwise, depending on the detection radius, the same issue with player thrown bottles happens.

Protonull committed:
Enabled Emerald Leveling

Players may now right click on an emerald to receive a bulk amount of xp, exactly nine times the amount of xp from a Bottle o' Enchanting.

Protonull commented:
Just added the feature we discussed in private, you may now right click emeralds to level up

Protonull committed:
Handles fishing xp

Protonull committed:
Disable emerald xp for lored emeralds

Protonull committed:
Handles Merchant and Breeding XP
